File: 6_tree/LCA问题/LCAManager.ts
function useLCAManager(n: number, adjMap: Map<number, Set<number>>, root = 0) {
  const depth = new Int32Array(n).fill(-1)
  const parent = new Int32Array(n).fill(-1)

  const _BITLEN = Math.floor(Math.log2(n)) + 1
  const _fa = Array.from<any, Int32Array>({ length: n }, () => new Int32Array(_BITLEN))

  _dfs(root, -1, 0)
  _initDp()

  /**
   * @description O(logn) 查询LCA
   */
  function queryLCA(root1: number, root2: number): number {
    if (depth[root1] < depth[root2]) [root1, root2] = [root2, root1]

    for (let bit = _BITLEN - 1; ~bit; bit--) {
      if (depth[_fa[root1][bit]] >= depth[root2]) {
        root1 = _fa[root1][bit]
      }
    }

    if (root1 === root2) return root1

    for (let bit = _BITLEN - 1; ~bit; bit--) {
      if (_fa[root1][bit] !== _fa[root2][bit]) {
        root1 = _fa[root1][bit]
        root2 = _fa[root2][bit]
      }
    }

    return _fa[root1][0]
  }

  function _dfs(cur: number, pre: number, dep: number): void {
    depth[cur] = dep
    parent[cur] = pre
    for (const next of adjMap.get(cur) ?? []) {
      if (next === pre) continue
      _dfs(next, cur, dep + 1)
    }
  }

  /**
   * @description O(nlogn) 初始化dp
   */
  function _initDp(): void {
    for (let i = 0; i < n; i++) _fa[i][0] = parent[i]
    for (let bit = 0; bit < _BITLEN - 1; bit++) {
      for (let i = 0; i < n; i++) {
        if (_fa[i][bit] === -1) _fa[i][bit + 1] = -1
        else _fa[i][bit + 1] = _fa[_fa[i][bit]][bit]
      }
    }
  }

  return {
    depth,
    parent,
    queryLCA,
  }
}

if (require.main === module) {
  function closestNode(n: number, edges: number[][], query: number[][]): number[] {
    const adjMap = new Map<number, Set<number>>()
    for (const [u, v] of edges) {
      adjMap.set(u, (adjMap.get(u) ?? new Set()).add(v))
      adjMap.set(v, (adjMap.get(v) ?? new Set()).add(u))
    }

    const { depth, queryLCA } = useLCAManager(n, adjMap)
    const res = []
    for (const [root1, root2, root3] of query) {
      const cands = [queryLCA(root1, root2), queryLCA(root2, root3), queryLCA(root1, root3)].sort(
        (a, b) => -(depth[a] - depth[b])
      )
      res.push(cands[0])
    }

    return res
  }
}

export { useLCAManager }

File: 6_tree/LCA问题/LCAManager.py
from collections import defaultdict
from math import floor, log2
from typing import DefaultDict, List, Set


class LCAManager:
    def __init__(self, n: int, adjMap: DefaultDict[int, Set[int]], root=0) -> None:
        """查询 LCA

        `nlogn` 预处理
        `logn`查询两点的LCA

        Args:
            n (int): 树节点编号 默认 0 ~ n-1
            adjMap (DefaultDict[int, Set[int]]): 树
            root (int, optional): 根节点 默认 0
        """
        self.depth = defaultdict(lambda: -1)
        self.parent = defaultdict(lambda: -1)
        self._BITLEN = floor(log2(n)) + 1
        self._MAX = n
        self._adjMap = adjMap
        self._dfs(root, -1, 0)
        self._fa = self._makeDp(self.parent)

    def queryLCA(self, root1: int, root2: int) -> int:
        """ `logn` 查询 """
        if self.depth[root1] < self.depth[root2]:
            root1, root2 = root2, root1

        for i in range(self._BITLEN - 1, -1, -1):
            if self.depth[self._fa[root1][i]] >= self.depth[root2]:
                root1 = self._fa[root1][i]

        if root1 == root2:
            return root1

        for i in range(self._BITLEN - 1, -1, -1):
            if self._fa[root1][i] != self._fa[root2][i]:
                root1 = self._fa[root1][i]
                root2 = self._fa[root2][i]

        return self._fa[root1][0]

    def queryDist(self, root1: int, root2: int) -> int:
        """查询树节点两点间距离"""
        return self.depth[root1] + self.depth[root2] - 2 * self.depth[self.queryLCA(root1, root2)]

    def _dfs(self, cur: int, pre: int, dep: int) -> None:
        """处理高度、父节点信息"""
        self.depth[cur], self.parent[cur] = dep, pre
        for next in self._adjMap[cur]:
            if next == pre:
                continue
            self._dfs(next, cur, dep + 1)

    def _makeDp(self, parent: DefaultDict[int, int]) -> List[List[int]]:
        """nlogn预处理"""
        dp = [[0] * self._BITLEN for _ in range(self._MAX)]
        for i in range(self._MAX):
            dp[i][0] = parent[i]
        for j in range(self._BITLEN - 1):
            for i in range(self._MAX):
                if dp[i][j] == -1:
                    dp[i][j + 1] = -1
                else:
                    dp[i][j + 1] = dp[dp[i][j]][j]
        return dp

